### PR TITLE
Resolve #51, use new asm! macro

### LINKED ISSUE #51: Use new asm! macro
https://github.com/Amanieu/rfcs/blob/inline-asm/text/0000-inline-asm.md

the `llvm_asm` macro is certainly not clear. Switch to this approach when we can. It isn't clear to me how much of this is done in: https://github.com/rust-lang/rust/pull/69171

### PR DESCRIPTION
**Pull request by maelgui**
Resolve issue #51 

We replaced all the `llvm_asm!` call with the new `asm!` macro, which implies:
- Use intel syntax by default (coherent with asm code used in the project)
- We tried to choose the best option (nomem, nostack, etc...) for each call
- "m" memory operand does not exist anymore, which implies passing reference address through a register

We tried to ensure the generated assembly stayed the same, using rust playground.

### CODE REVIEW

**Comment by ALSchwalm** [comment]
Awesome! I'll take a look at this tomorrow. Before merger though, I'd like some of these commits squashed, e.g. 
92d1672 looks like a fix for an earlier commit, and f385de6 has the same commit message as c9e846f. But at a glance this looks really solid.

**Comment by maelgui** [comment]
I rewrote the history, and squashed the two or three commits that were the same.

File: mythril/src/error.rs [hunk 1]
```diff
@@ -52,11 +52,12 @@ pub fn check_vm_insruction(rflags: u64, error: String) -> Result<()> {
...
             let value: u64;
-            llvm_asm!("vmread %rax, %rdx;"
-                      : "={rdx}"(value)
-                      : "{rax}"(vmcs::VmcsField::VmInstructionError as u64)
-                      : "rflags"
-                      : "volatile");
+            asm!(
+                "vmread rdx, rax",
+                in("rax") vmcs::VmcsField::VmInstructionError as u64,
+                out("rdx") value,
```
Comment (ALSchwalm) [inline mythril/src/error.rs:58]: Doesn't this need something like `lateout("rflags") _` to ensure that the compile is aware that rflags has been clobbered by this asm?

File: mythril/src/logger.rs [hunk 2]
```diff
@@ -83,11 +83,14 @@ pub unsafe fn raw_write_console(s: impl AsRef<str>) {
...
-         :"{rdx}"(0x3f8), "{rcx}"(len as u64), "{rsi}"(ptr as u64)
-         : "rflags", "rsi"
-         : "volatile");
+    asm!(
+        "cld",
+        "rep outsb",
+        in("rdx") 0x3f8,
+        in("rcx") len as u64,
+        inout("rsi") ptr as u64 => _,
+        options(nostack)
```
Comment (ALSchwalm) [inline mythril/src/logger.rs:92]: Again, I think we need to indicate that rflags has been clobbered by cld

**Comment by ALSchwalm** [comment]
Ok, reading through these changes it looks good to me. I had one concern, which is that in a few places we no longer express that the `rflags` register is clobbered by the asm. However, it appears that is not possible to specify, as it is implied unless `preserves_flags` is set (per https://github.com/rust-lang/project-inline-asm/issues/5). Does that align with your understanding of things? I'm not that familiar with the new `asm` stuff (but it certainly seems nicer to use than `llvm_asm`).

Thanks again for working on this.

**Comment by ALSchwalm** [comment]
I'm going to go ahead and merge this. Thanks for the contribution!

**Comment by paulcacheux** [comment]
Thanks for merging it. What you figured about `preserves_flags` matches our understanding of the new macro. It seems that the new options of the `asm!` macro are defined as opposites of the old `llvm_asm` ones such as no options is nearly always at least correct and then you can optimize by providing more information (like `nomem`, `nostack`, `pure`, `preserves_flags`, etc).

**Pull request merged**